### ISSUE TITLE
Cannot backward twice without calling zeroGradParameters in between

### ISSUE THREAD

**Issue by jonathanasdf** [torch/nn#792]
I was under the impression that gradients should be accumulated, and there should not be a need to call zeroGradParameters before every call to backward. But, here's a simple example where that seems not to be true...

```
require 'nn'
local m = nn.Linear(1, 1)
local p, gp = m:getParameters()
local c = nn.MSECriterion()
local t = torch.Tensor{0}
local i = torch.rand(1, 1)
local o = m:forward(i)
c:forward(o, t)
local g = c:backward(o, t)
m:zeroGradParameters()
m:backward(i, g)
print(gp:abs():sum())
--m:zeroGradParameters()
m:backward(i, g)
print(gp:abs():sum())
```

If you run it a couple times, you will see that sometimes the second magnitude is twice the first one, which is the expected correct behavior, but other times the second magnitude becomes 0. Uncommenting the zeroGradParameters line makes everything behave as expected.

Is this a bug, or should the docs be updated to mention that zeroGradParameters must be called before each call to backward (in which case why not add an assert inside backward?)

**Comment by fmassa** [comment]
I managed to reproduce this behaviour, and I should say that this looks quite weird. In the case where it doesn't work, by some crazy reason the `gradBias` (and `gradWeight`) have the wrong sign, making this alternate behaviour. I tried to run by hand the commands executed inside the function call (in `accGradParameters`), and it was all working as expected.

I have no idea at the moment where the problem is...

**Comment by albanD** [comment]
`gp:abs()` is inplace ;)

**Comment by fmassa** [comment]
Haha ! That's it ! Thanks @albanD, I oversaw that !

**Comment by fmassa** [comment]
@albanD in your edited example, for completeness you forgot to add the out of place version of `abs` in the last line

**Comment by albanD** [comment]
@fmassa thanks, I edited the code.

**Comment by jonathanasdf** [comment]
I feel so silly now.. Closing
